source/h1_encoder.c

```diff
@@ -712,116 +443,308 @@ int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct
+    const size_t prev_len = dst->len;
+    int err = aws_input_stream_read(stream, dst);
+    const size_t amount_read = dst->len - prev_len;
+
+    if (err) {
+        ENCODER_LOGF(
+            ERROR,
+            encoder,
+            "Failed to read body stream, error %d (%s)",
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+
+        return AWS_OP_ERR;
+    }
+
+    /* Increment progress_bytes, and make sure we haven't written too much */
+    int add_err = aws_add_u64_checked(encoder->progress_bytes, amount_read, &encoder->progress_bytes);
+    if (add_err || encoder->progress_bytes > total_length) {
+        ENCODER_LOGF(ERROR, encoder, "Body stream has exceeded expected length: %" PRIu64, total_length);
+        return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+    }
+
+    ENCODER_LOGF(
+        TRACE,
+        encoder,
+        "Sending %zu bytes of body, progress: %" PRIu64 "/%" PRIu64,
+        amount_read,
+        encoder->progress_bytes,
+        total_length);
+
+    /* Return if we're done sending stream */
+    if (encoder->progress_bytes == encoder->message->content_length) {
+        *out_done = true;
+        return AWS_OP_SUCCESS;
+    }
+
+    /* Return if stream failed to write anything. Maybe the data isn't ready yet. */
+    if (amount_read == 0) {
+        /* Ensure we're not at end-of-stream too early */
+        struct aws_stream_status status;
+        err = aws_input_stream_get_status(encoder->message->body, &status);
+        if (err) {
             ENCODER_LOGF(
-                ERROR,
+                TRACE,
                 encoder,
-                "Failed to encode chunk, error %d (%s)",
+                "Failed to query body stream status, error %d (%s)",
                 aws_last_error(),
                 aws_error_name(aws_last_error()));
+
             return AWS_OP_ERR;
         }
+        if (status.is_end_of_stream) {
+            ENCODER_LOGF(
+                ERROR,
+                encoder,
+                "Reached end of body stream but sent less than declared length %" PRIu64 "/%" PRIu64,
+                encoder->progress_bytes,
+                total_length);
+            return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+        }
     }
-    /* return success to schedule the rest of the transmission on the next message */
+
+    /* Not done streaming data out yet */
     return AWS_OP_SUCCESS;
 }
 
-static int s_h1_encoder_process_body(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder) {
+/* A state function should:
+ * - Raise an error only if unrecoverable error occurs.
+ * - `return s_switch_state(...)` to switch states.
+ * - `return AWS_OP_SUCCESS` if it can't progress any further (waiting for more
+ *    space to write into, waiting for more chunks, etc). */
+typedef int encoder_state_fn(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst);
```

Comment:
From here on down is the new state machine. The logic of how bytes are copied is basically the same as before. But now each state is its own function, and there's one super simple state machine loop that drives (loop is at bottom of file)